internal/simplekvtest/keyvalue.go

```diff
@@ -0,0 +1,246 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package simplekvtest
+
+import (
+	"time"
+
+	"golang.org/x/net/context"
+	gc "gopkg.in/check.v1"
+	errgo "gopkg.in/errgo.v1"
+
+	"github.com/juju/simplekv"
+)
+
+// KeyValueSuite contains a set of tests for KeyValueStore implementations. The
+// NewStore parameter must be set before calling SetUpTest.
+type KeyValueSuite struct {
+	NewStore func(name string) (simplekv.Store, error)
+}
+
+func (s *KeyValueSuite) SetUpSuite(c *gc.C) {}
+
+func (s *KeyValueSuite) TearDownSuite(c *gc.C) {}
+
+func (s *KeyValueSuite) SetUpTest(c *gc.C) {}
+
+func (s *KeyValueSuite) TearDownTest(c *gc.C) {}
+
+func (s *KeyValueSuite) TestSet(c *gc.C) {
+	kv, ctx, close := s.newStore(c, "kv")
+	defer close()
+
+	err := kv.Set(ctx, "test-key", []byte("test-value"), time.Time{})
+	c.Assert(err, gc.Equals, nil)
+
+	result, err := kv.Get(ctx, "test-key")
+	c.Assert(err, gc.Equals, nil)
+	c.Assert(string(result), gc.Equals, "test-value")
+
+	// Try again with an existing record, which might trigger different behavior.
+	err = kv.Set(ctx, "test-key", []byte("test-value-2"), time.Time{})
+	c.Assert(err, gc.Equals, nil)
+
+	result, err = kv.Get(ctx, "test-key")
+	c.Assert(err, gc.Equals, nil)
+	c.Assert(string(result), gc.Equals, "test-value-2")
+}
+
+func (s *KeyValueSuite) TestGetNotFound(c *gc.C) {
+	kv, ctx, close := s.newStore(c, "kv")
+	defer close()
+
+	_, err := kv.Get(ctx, "test-not-there-key")
+	c.Assert(errgo.Cause(err), gc.Equals, simplekv.ErrNotFound)
+	c.Assert(err, gc.ErrorMatches, "key test-not-there-key not found")
+}
+
+func (s *KeyValueSuite) TestSetKeyOnce(c *gc.C) {
+	kv, ctx, close := s.newStore(c, "kv")
+	defer close()
+
+	err := simplekv.SetKeyOnce(ctx, kv, "test-key", []byte("test-value"), time.Time{})
+	c.Assert(err, gc.Equals, nil)
+
+	result, err := kv.Get(ctx, "test-key")
+	c.Assert(err, gc.Equals, nil)
+	c.Assert(string(result), gc.Equals, "test-value")
+}
+
+func (s *KeyValueSuite) TestSetKeyOnceDuplicate(c *gc.C) {
+	kv, ctx, close := s.newStore(c, "kv")
+	defer close()
+
+	err := simplekv.SetKeyOnce(ctx, kv, "test-key", []byte("test-value"), time.Time{})
+	c.Assert(err, gc.Equals, nil)
+
+	err = simplekv.SetKeyOnce(ctx, kv, "test-key", []byte("test-value"), time.Time{})
+	c.Assert(errgo.Cause(err), gc.Equals, simplekv.ErrDuplicateKey)
+	c.Assert(err, gc.ErrorMatches, "key test-key already exists")
+}
+
+func (s *KeyValueSuite) TestTwoStoresForTheSameIDPCommunicate(c *gc.C) {
+	kv1, ctx1, close1 := s.newStore(c, "kv")
```

Comment:
I suspect this functionality is a little too convoluted for a generic and simple kv store implementation. it adds a bit too much complexity for the simple use case and can be easily added in use cases where it is needed. I would drop this functionality and just create stores based on a passed in mongo collection, or sql table name, or whatever.

Reply:
done